### PR TITLE
top-level: Duplicate overlaying unless stdenvOverrides comes last

### LINKED ISSUE #34086: overrideAttrs and overlays don't play well together (overrides get invoked twice)
## Issue description

@michaelpj @Ericson2314 If you do an overrideAttrs in a overlay, it gets applied twice.

### Steps to reproduce

Here is an example.  Say we wanted to overlay gnupatch to pass "--disable-merge" to the configure routine.  We would write the following overlay (i.e., put this in *~/.config/nixpkgs/overlays/gnupatch.nix*)

```nix
self: super:  {
  gnupatch = super.gnupatch.overrideAttrs (old: {
    configureFlags = old.configureFlags or [] ++ [ "--disable-merge" ];
  });
}
```

But look what happens when we evaluate it

```bash
nix-instantiate --eval --expr '(import <nixpkgs> {}).gnupatch.configureFlags'
```
```nix
[ "--disable-merge" "--disable-merge" ]
```

Adding `trace` calls shows that the close passed to `overlayAttrs` get invoked twice even though the overlay itself only gets invoked once.

```nix
builtins.trace "overlay invoked" (
  self: super: {
    gnupatch = builtins.trace "gnupatch invoked" (
      super.gnupatch.overrideAttrs (builtins.trace "override invoked" (
        old: {
          configureFlags = builtins.trace (old.configureFlags or []) (
            old.configureFlags or [] ++ [ "--disable-merge" ]
          );
        }
      ))
    );
  }
)
```

```bash
nix-instantiate --eval --expr '(import <nixpkgs> {}).gnupatch.configureFlags'
```
```nix
trace: overlay invoked
trace: gnupatch invoked
trace: gnupatch invoked
trace: override invoked
trace: override invoked
trace: [ ]
trace: [ "--disable-merge" ]
[ "--disable-merge" "--disable-merge" ]
```

It is almost as if `overrideAttrs` has somehow pulled in the final desired `gnupatch` expression (i.e., the already overriden one) and then applied the override to it again (giving it twice applied) despite the fact that it was taken from the prior `gnupatch` expression (i.e., `super.gnupatch.overrideAttr`).

Note that this isn't a simple case of `self` and `super` being mixed up.  Switching them gives an infinite recursion error as expected.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

> - system: `"x86_64-linux"`
> - host os: `Linux 4.9.72, NixOS, 17.09.2484.3aec59c99ff (Hummingbird)`
> - multi-user?: `yes`
> - sandbox: `no`
> - version: `nix-env (Nix) 1.12pre5788_e3013543`
> - channels(root): `"nixos-17.09.2484.3aec59c99ff"`
> - channels(tyson): `""`
> - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

### PR DESCRIPTION
**Pull request by twhitehead**
The stdenvOverrides overlay is used to bring packages forward during bootstrapping via stdenv.overrides.  These packages have already had the overlays applied to them in the previous boostrapping stage.  If stdenvOverrides is not last in the overlays stack, all remaining overlays will windup being applied again to these packages.

closes #34086

###### Motivation for this change

See #34086 and above description.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]
Great sluething! I'll wait for @nbp's final approval but looks good to me.

**Review by nbp — APPROVED** [review]
https://github.com/NixOS/nixpkgs/issues/34086#issuecomment-360991748:
> Possibly stdenvOverrides should always be at the end of overlays list in stage.nix. If it is only used to bring packages forward, this could be valid as previous packages will already have the overlays applied to them. I don't know if this is correct, but I'll submit a pull request.

If this work, this means that stdenvOverrides do not take into account the previous overlays applied in the `super` stages, which in it-self should probably be considered as a bug.

Also, do we want to even apply overlays to stdenv packages? From my understanding, doing so implies that we would be recompiling everything.

I will approve this commit as this work-around an issue, but be aware that this should not be the definitive solution.

The best solution would be to move stdenvOverrides into all-packages.nix  The fact that we override the previous stages should also move to Nixpkgs, by making stages as a form of nesting.
```nix
{
  stdenvLinux = {
    prevStage = {
      prevStage = {
      };
    };
  }
}
```

**Comment by twhitehead** [comment]
Not sure I fully understand the "move `stdenvOverrides` into _all-packages.nix_" comment.  A couple of comments that may be relevant though

1. The stdenv override varies by boot stage so at least what is overridden needs to be declared in the stdenv stage files (i.e., I don't  think it can be  entirely  moved into _all-packages.nix_).
2.  There is an existing mechanism to only apply an override to the last stage.  This is done via the `allowCustoOverrides` attribute.  Currently only the `configOverrides` uses this.

I would also note that we would appreciate not having the overlay functionality too artificially limited as we are looking to use it to overlay (potentially quite deeply) [HPC packages and tweaks](https://git.computecanada.ca/nix/ccpkgs) (ignore the README, I have to rewrite it) onto the standard nix package set

**Comment by Ericson2314** [comment]
Merged as a good stop gap. But let's continue discussing how to make things better.

**Pull request merged**
